### PR TITLE
fix(learning, macos): isolate LightGBM training in a subprocess to prevent daemon SIGSEGV (dual-libomp OpenMP conflict)

### PR DESCRIPTION
**Pull request by barrygfox**
## Summary

  On macOS, clicking **"Train model now"** in the dashboard Brain pane
  (`POST /api/learning/retrain`) — and calling `POST /api/v3/learning/consolidate`
  — hard-crashes the unified daemon. The HTTP call returns an empty reply and the
  daemon PID changes (it gets auto-restarted). No Python traceback is logged,
  because the process dies on a native signal.

  This PR runs all LightGBM training in an isolated subprocess so it can never
  crash the daemon, mirroring the existing `embedding_worker` / `reranker_worker`
  isolation pattern.

  ## Root cause

  The daemon serves the HTTP API **in-process** and, by then, has PyTorch's
  OpenMP runtime (`torch/lib/libomp.dylib`) loaded with **warm worker threads**
  (the cross-encoder reranker + embedding warm-up at startup). When a request
  handler imports `lightgbm` and trains, LightGBM loads a **second** OpenMP
  runtime — the Homebrew `libomp.dylib` that `lib_lightgbm.dylib` links against —
  and forks its own OMP worker pool. Two live OpenMP runtimes in one process
  corrupt the shared `__kmp` state, and a worker thread segfaults:

  ```
  Termination: SIGSEGV (EXC_BAD_ACCESS)
  Faulting thread: libomp.dylib __kmp_launch_worker
                   → __kmp_fork_barrier → __kmp_suspend_initialize_thread
  ```
  (`~/Library/Logs/DiagnosticReports/Python-*.ips`)

  ### An OpenMP conflict, not a model load path issue (also suggested as a trigger)

  Experiments on my machine (Python 3.14 / macOS ARM64 / LightGBM 4.6.0)
  provide a strong candidate root cause:

  | Condition | `model_file=` | `model_str=` | Outcome |
  |---|---|---|---|
  | Clean process, **no torch** | ✅ | ✅ | not a standalone LightGBM bug |
  | **torch warm**, default threads | 💥 SIGSEGV | 💥 SIGSEGV | crashes at the first booster/training op |
  | torch warm + `OMP_NUM_THREADS=1` | ✅ | ✅ | no OMP thread-forking, no issue |

  The likely issue is **torch's OpenMP pool being warm when LightGBM first forks
  its own pool**. Import **order** matters too: lightgbm-first then
  torch is safe; torch-warm-first then lightgbm crashes.

  The test suite mostly survives because `conftest.py` sets
  `OMP_NUM_THREADS=1` (serial OpenMP, no parallel fork). The **daemon** does not have this setting and runs parallel OpenMP, so it (daemon) crashes in production.

  ## Reproduction

  No SuperLocalMemory code involved. Requires `torch`,
  `lightgbm`, and `numpy` in one environment (Python 3.14 / macOS Apple Silicon /
  LightGBM 4.6.0):

  ```python
  """torch-warm process + LightGBM => SIGSEGV (macOS / Apple Silicon).
  Dies with SIGSEGV and no traceback on affected platforms; prints DONE elsewhere."""
  import sys
  def step(m): sys.stderr.write(m + "\n"); sys.stderr.flush()

  import numpy as np
  import torch

  step("1) warming torch's OpenMP pool (parallel matmul)")
  _ = (torch.rand(1500, 1500) @ torch.rand(1500, 1500)).sum().item()

  step("2) importing lightgbm + training (forks LightGBM's own OpenMP workers)")
  import lightgbm as lgb
  X = np.random.RandomState(0).rand(300, 6); y = (X[:, 0] > 0.5).astype(int)
  booster = lgb.train({"objective": "binary", "verbosity": -1},
                      lgb.Dataset(X, label=y), num_boost_round=8)

  step("3) loading the booster back")
  lgb.Booster(model_str=booster.model_to_string())

  step("DONE — no crash on this platform")
  ```

  Observed:

  ```console
  $ python repro.py
  1) warming torch's OpenMP pool (parallel matmul)
  2) importing lightgbm + training (forks LightGBM's own OpenMP workers)
  [1]    <pid> segmentation fault  python repro.py          # exit 139, no traceback

  $ OMP_NUM_THREADS=1 python repro.py
  1) warming torch's OpenMP pool (parallel matmul)
  2) importing lightgbm + training (forks LightGBM's own OpenMP workers)
  3) loading the booster back
  DONE — no crash on this platform                          # exit 0
  ```

  The crash fires at step 2 — the first LightGBM OpenMP fork after torch's pool is
  warm — the daemon's cold-start retrain path. Swap the order
  (import + train LightGBM *before* warming torch) and it also runs clean,
  confirming the issue is two live OpenMP runtimes in one process rather than
  SuperLocalMemory code. The daemon hits this because the reranker/embedding
  warm-up makes torch's OpenMP pool warm before any training request arrives.

  ## Reachable crash sites

  - `POST /api/learning/retrain` → `_retrain_ranker_impl`
  - `POST /api/v3/learning/consolidate` → `ConsolidationWorker.run()` step 5
    (online `_run_shadow_cycle` or legacy cold-start)

  Not affected: the maintenance endpoints / MCP `run_maintenance` (pattern mining
  only, no LightGBM), the CLI `slm consolidate --cognitive` (CCQ pipeline), and
  the background schedulers (none train LightGBM) — so there is no silent
  on-a-timer crash; every crash requires an explicit user/API call.

  ## The proposed fix

  A new module `learning/lightgbm_subprocess.py` exposing `run_retrain_isolated()`
  and `run_consolidation_isolated()`. They spawn a fresh process via a
  `python -c "import lightgbm; from ... import main; ..."` bootstrap that imports
  `lightgbm` **before** the `superlocalmemory` package (which transitively imports
  torch). This ordering is critical — `python -m superlocalmemory…` would run
  the package `__init__` (torch) first and crash. The child never runs a
  torch tensor op, so torch's OMP pool stays dormant and LightGBM's runtime is the
  only active pool, thereby avoiding the crash scenario.

  - The child emits a single JSON verdict line; the parent parses it and **never
    raises** — a native child crash is reported as an error, keeping the daemon up.
  - On success the daemon calls `model_cache.invalidate(profile)` so the freshly
    trained model is reloaded.
  - Both endpoints call the helpers via `run_in_threadpool` to avoid blocking the
    event loop.

  ## Alternatives considered

  - **`OMP_NUM_THREADS=1` in the daemon** — would prevent the crash, but serializes
    *all* torch/sklearn/LightGBM parallelism in the daemon (hurts reranker and
    embedding throughput). Subprocess isolation keeps the daemon parallel and
    isolates only training.
  - **Loading boosters via `model_file=` (temp file)** — does **not** fix it; under
    torch-warm, `model_file=` segfaults too (see table).
  - **Unifying the libomp copies via `install_name_tool`** — fragile; `pip` reinstall
    re-runs delocate and resets install names.

  ## Testing

  - New `tests/test_learning/test_lightgbm_subprocess.py`: lightgbm-first bootstrap
    ordering, flag forwarding, and crash/timeout/verdict handling for both tasks.
  - Manual end-to-end: the daemon PID is **stable** across both endpoints
    (HTTP 200) where it previously crashed.

  ## Notes

  - `/api/learning/retrain` continues to use the (deprecated) cold-start
    `_retrain_ranker_impl`, because the button's "train + activate now" semantics
    have no non-deprecated replacement yet (the online `_run_shadow_cycle` only
    persists a *candidate*; promotion is deferred to the shadow-router). This PR
    centralizes that deprecated reference into a single swappable function
    (`lightgbm_subprocess._retrain`) so removing the legacy path later is a
    one-line change.
 - Platform: the SIGSEGV is observed on macOS / Apple Silicon (wheel-packaging quirk — multiple vendored libomp copies). The fix isolates training for all OS; on Linux/Windows the cost is an infrequent subprocess spawn but the proposed fix yields a "training can't crash the daemon" guarantee across all OS.
 - 
  ## Files changed

  - `src/superlocalmemory/learning/lightgbm_subprocess.py` (new)
  - `src/superlocalmemory/server/routes/learning.py`
  - `src/superlocalmemory/server/routes/v3_api.py`
  - `tests/test_learning/test_lightgbm_subprocess.py` (new)

### CODE REVIEW

**Comment by varun369** [comment]
Merged ✅ — thanks for the detailed write-up and the clean repro.

We verified independently on the same environment as the report (macOS Apple Silicon, Python 3.14.4, torch 2.11.0, lightgbm 4.6.0, two `libomp.dylib` copies present: `torch/lib` + `sklearn/.dylibs`):

- **Bug reproduced unmasked** — torch-warm process + in-process LightGBM train → `exit 139` SIGSEGV (the test suite's `conftest.py OMP_NUM_THREADS=1` was correctly identified as what hides it).
- **Fix mechanism confirmed** — torch-warm parent + isolated lightgbm-first child → child exits 0 with a clean JSON verdict, parent survives. Daemon would stay up.
- **Defensive contract holds** — `run_retrain_isolated` on a bad path returns `{trained: False, error: ...}` rather than raising.
- **Symbols resolve** — `ranker_retrain_legacy._retrain_ranker_impl`, `model_cache.invalidate`, `ConsolidationWorker` all present; modified routes + new module import clean.
- Test file: 9/9 passing.

Centralizing the deprecated cold-start reference into a single swappable `_retrain` is a nice touch for the eventual legacy removal.

Follow-up: this ships to users in the next patch release (3.5.6) — npm + PyPI bumped together via the `v*` tag. The merge alone doesn't update installed daemons (non-editable site-packages install).
